Classes basics

DavydovDenys opened the pull request.

DavydovDenys committed:
Add Logic for Transport

DavydovDenys committed:
Add constants

DavydovDenys committed:
Add logic for Bike

DavydovDenys committed:
Add logic for Car

DavydovDenys committed:
Add logic for DeliveryService

DavydovDenys committed:
Add VehicleAvailabilities to controll available transport

DavydovDenys committed:
Remove comments

dmgrinko1 commented on bike.rb:
```diff
@@ -1,2 +1,15 @@
+    @speed = 10
+    @max_weight = 10
+    @max_distance = 30
+  end
```
All this values must comes from Constant. Something like 
CARS = {
    max_weight: 100,
    speed: 50,
     .....
  }.freeze

DavydovDenys replied:
done

dmgrinko1 commented on bike.rb:
```diff
@@ -1,2 +1,15 @@
+
+  def available
+    Car.availability.bikes
+  end
```
what this method was created for?

DavydovDenys replied:
removed

dmgrinko1 commented on car.rb:
```diff
@@ -1,2 +1,14 @@
+  def initialize
+    @speed = 50
+    @max_weight = 100
+  end
```
All this values must comes from Constant.

DavydovDenys replied:
done

dmgrinko1 commented on car.rb:
```diff
@@ -1,2 +1,14 @@
+  def available
+    Car.availability.cars
+  end
+end
```
what this method was created for?

DavydovDenys replied:
removed

dmgrinko1 commented on constants.rb:
```diff
@@ -1,2 +1,4 @@
\ No newline at end of file
+  CARS = 20
+  BIKES = 30
+end
```
20 and 30 of what? quantity, volume, etc?

DavydovDenys replied:
done

dmgrinko1 commented on transport.rb:
```diff
@@ -1,2 +1,18 @@
\ No newline at end of file
+  extend VehicleAvailabilities
+  attr_accessor :max_weight, :speed, :available
+
```
according to the task :available (boolean) attribute must be here in the inizialize method, cause have to work with both vehicle (Car, Bike)

dmgrinko1 commented on transport.rb:
```diff
@@ -1,2 +1,18 @@
+    (distance / speed.to_f).round(2)
+  end
+
+  def delivery_speed_per_kg
```
its not good idea to use units in the method name, cause they can be different
lets name it delivery_speed_by_weight

dmgrinko1 commented on transport.rb:
```diff
@@ -1,2 +1,18 @@
+    (speed.to_f / max_weight).round(2)
+  end
+
+  def available_vehicle
```
I don`t see any place where you using this method

dmgrinko1 commented on vehicle_availabilities.rb:
```diff
@@ -0,0 +1,10 @@
+  AVAILABILITY = Availability.new(Constants::CARS, Constants::BIKES)
+
+  def availability
+     AVAILABILITY
```
All this availability functional within module module Constants and module VehicleAvailabilities
Does not make sense. Cause quantity of Bike and Cars can not be hardcocded. It depends of how mutch Car or Bike objects you created.
In your case logic is wronk. Cause I can create something like 1000.times {Bike.new} but still will have BIKES = 30

DavydovDenys replied:
removed

dmgrinko1 commented on delivery_service.rb:
```diff
@@ -1,2 +1,37 @@
+    @cars = DeliveryService.availability
+  end
+
+  def vehicle(weight, distance)
```
this method should be named like create_delivery or something more informative

DavydovDenys replied:
done

dmgrinko1 commented on delivery_service.rb:
```diff
@@ -1,2 +1,37 @@
\ No newline at end of file
+  extend VehicleAvailabilities
+
+  def initialize
```
you can use something like that to create vehicle park array 
def initialize
    @bikes = Array.new('array size here')) { Bike.new }
    @cars = Array.new('array size here')) { Car.new }
  end

DavydovDenys replied:
done

dmgrinko1 commented on delivery_service.rb:
```diff
@@ -1,2 +1,37 @@
+  extend VehicleAvailabilities
+
+  def initialize
+    @cars = DeliveryService.availability
```
we don`t need hard coded avability count
also variable name @cars is not correct cause DeliveryService.availability provides bikes and cars

DavydovDenys replied:
removed

dmgrinko1 commented on delivery_service.rb:
```diff
@@ -1,2 +1,37 @@
+  end
+
+  def vehicle(weight, distance)
+    bikes = bikes()
```
we don`t need this vabialbe

DavydovDenys replied:
removed

DavydovDenys committed:
Add changes according to comments

DavydovDenys committed:
Add RSpec

Add RSpec and specs

DavydovDenys committed:
Add quantity

Get quantity for cars, bikes from Constants

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -0,0 +1,11 @@
+  def delivery_speed_by_weight
+    (speed.to_f / max_weight).round(2)
+  end
+end
```
I don`t see here comparable method from task requirement?
'расширить класс Transport чтобы объекты этого класса можно было сравнивать по скорости доставки 1 кг посылки'

dmgrinko1 replied:
Class Transport does not have required attributes from task
'создать базовый класс Transport c 3 атрибутами(max_weight, speed и available)'
your Transport object => #<Transport:0x00007fd7ae032ae0>

dmgrinko1 replied:
Bike, Car, Transport have same attributes like :max_weight, :speed , :available. So you have to initialize it in Transport class instead of Bike and Car. Also it will wif previous comment.
Here is exaple you need from Bike class
def initialize
    super(Constants::BIKES[:max_weight], Constants::BIKES[:speed], true)
    @max_distance = Constants::BIKES[:max_distance]
  end

DavydovDenys replied:
Add comparable method, add required attributes.

DavydovDenys replied:
Change initialize method of Bike, Car with super method. Would you explain how to implement a comparable method to compare two objects? If my method compare_delivery_speed_by_weight_with is useless.

dmgrinko1 replied:
@DavydovDenys you can do it via comparable module. It was explained in task tips.
https://ruby-doc.org/core-1.8.6/Comparable.html

DavydovDenys committed:
Update Transport

Add initialize
Add compare_delivery_speed_by_weight_with

DavydovDenys committed:
Add super() to initialize of Bike, Car

DavydovDenys committed:
Add Comparable module to Transport

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -0,0 +1,24 @@
+  include Comparable
+
+  attr_accessor :max_weight, :speed, :available
+  attr_reader :delivery_speed_by_weight
```
you applied attr_reader for method instead of variable, it will work without it. remove it please

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -0,0 +1,24 @@
+class Transport
+  include Comparable
+
+  attr_accessor :max_weight, :speed, :available
```
:max_weight and :speed comes from constants and should not be editable, please use attr_reader here.

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -0,0 +1,24 @@
+
+  def initialize(max_weight, speed, available)
+    @max_weight = max_weight
+    @speed = speed
```
lets to speed.to_f in inizialize method instead of using it twice (delivery_time and delivery_speed_by_weight methods)

DavydovDenys committed:
Add changes according to comments

Add attr_reader to :max_weight, :speed
Add speed.to_f in initialize of Transport class

DavydovDenys commented:
Thank you for your comments! I finally understand the difference between super and super() and find out about Comparable module.
